contracts/verifying/VerifyingSingletonPaymasterV2.sol

```diff
@@ -1,9 +1,358 @@
+     * @param withdrawAddress The address to which the gas tokens should be transferred.
+     * @param amount The amount of gas tokens to withdraw.
+     */
+    function withdrawTo(
+        address payable withdrawAddress,
+        uint256 amount
+    ) public override nonReentrant {
+        if (withdrawAddress == address(0)) revert CanNotWithdrawToZeroAddress();
+        uint256 currentBalance = paymasterIdBalances[msg.sender];
+        if (amount > currentBalance)
+            revert InsufficientBalance(amount, currentBalance);
+        paymasterIdBalances[msg.sender] =
+            paymasterIdBalances[msg.sender] -
+            amount;
+        entryPoint.withdrawTo(withdrawAddress, amount);
+        emit GasWithdrawn(msg.sender, withdrawAddress, amount);
+    }
+
+    /**
+     * @dev Set a new verifying signer address.
+     * Can only be called by the owner of the contract.
+     * @param _newVerifyingSigner The new address to be set as the verifying signer.
+     * @notice If _newVerifyingSigner is set to zero address, it will revert with an error.
+     * After setting the new signer address, it will emit an event VerifyingSignerChanged.
+     */
+    function setSigner(
+        address _newVerifyingSigner
+    ) external payable override onlyOwner {
+        if (_newVerifyingSigner == address(0))
+            revert VerifyingSignerCannotBeZero();
+        address oldSigner = verifyingSigner;
+        assembly {
+            sstore(verifyingSigner.slot, _newVerifyingSigner)
+        }
+        emit VerifyingSignerChanged(oldSigner, _newVerifyingSigner, msg.sender);
+    }
+
+    function setUnaccountedEPGasOverhead(
+        uint256 value
+    ) external payable onlyOwner {
+        uint256 oldValue = unaccountedEPGasOverhead;
+        unaccountedEPGasOverhead = value;
+        emit EPGasOverheadChanged(oldValue, value);
+    }
+
+    function setFixedPriceMarkup(uint32 _markup) external payable onlyOwner {
+        require(_markup <= PRICE_DENOMINATOR * 2, "markup too high");
+        uint32 oldValue = fixedPriceMarkup;
+        fixedPriceMarkup = _markup;
+        emit FixedPriceMarkupChanged(oldValue, _markup);
+    }
+
+    /**
+     * @dev This method is called by the off-chain service, to sign the request.
+     * It is called on-chain from the validatePaymasterUserOp, to validate the signature.
+     * @notice That this signature covers all fields of the UserOperation, except the "paymasterAndData",
+     * which will carry the signature itself.
+     * @return hash we're going to sign off-chain (and validate on-chain)
+     */
+    function getHash(
+        UserOperation calldata userOp,
+        address paymasterId,
+        uint48 validUntil,
+        uint48 validAfter,
+        uint32 priceMarkup
+    ) public view returns (bytes32) {
+        //can't use userOp.hash(), since it contains also the paymasterAndData itself.
+        return
+            keccak256(
+                abi.encode(
```

Comment:
since all elements are of constant size, use `abi.encodePacked()` to save some gas

Reply:
or keep abi.encode() but remove intermediate keccak256

Reply:
I could remove intermediate keccak256 for initCode and calldata.

Reply:
btw in all the references I have seen it is like this
https://github.com/eth-infinitism/account-abstraction/blob/develop/contracts/samples/VerifyingPaymaster.sol#L42

Reply:
can remove intermediate keccak256